### ISSUE TITLE
Typescript 5 decorators

### ISSUE THREAD

**Issue by pokatomnik** [denoland/deno#19160]
Release [1.32.0](https://github.com/denoland/deno/releases/tag/v1.32.0) comes with Typescript 5.0.2. But, as per notes in [this](https://github.com/denoland/deno/pull/18294) PR I see no modern (stage 3) decorators support.
As far as I understand, the problem is about SWC.
But SWC [supports](https://swc.rs/docs/configuration/compilation#jsctransformdecoratorversion) them in the current version.
Could you please clarify, what version of SWC used in the latest Deno release? And will they be supported some day at all?

**Comment by pokatomnik** [comment]
[Here is](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAGTgczQUwE4B4BQiiAKgJJgDOUAhpJopgB5SZgAmFicARgFabQANIRIBBbGhABbVlE5MW7TuOzUAnrnABrMHADuYAHzCixAEqYoIbEgWsOibboP5DAChFxsMNDDDUAGwBZSwALODYALkQ3KFCYCmiySho6QUQAOizqCUSxCWlZCgBKRABeQxILKxsTRAgEFmZogGEA6goKELiIgBEBL2ooLxbGhQIiU3IqWghMOqJY+LzkmbTM7Nyk8UkZMDlSiqrLazARQ3xSgG8RbBObRFBIWAQYuISk6dS59KyMnLQKx2hX2JSS1VOiBuk3qCAocACmAyAXQbgARAAJTABFGPbBwKSIFEYHBo4oAbhERDuNSQXh8fkC3XCbAyEECASWCV+m0BFJEAF9KQL8PgIO1OiRMFQoSIAAKoEnYEQABxA3ACMAgiBYVDc1GiYGk3Bw6W40So3jAaGuVNhlARSOJbgABsC9nJEDlMNEACRXagC9L+7gCl38ogikViuFQHXSuNlRBgTB6KV6-m6qAZLNuACM6TRACYyZSgA) a simple example that is perfectly working in ts-node, but not working in Deno.
I believe Deno recognizes these decorators as legacy ones and fails (typically the are being ignored)

**Comment by jiawei397** [comment]
I am also following this progress. I have a framework oak_nest heavily relies on decorators, and if there are changes, it may be necessary to modify the code accordingly to support new decorators.

**Comment by nberlette** [comment]
Hey guys, any progress update on this feature's implementation? It's a real bummer to have all of TypeScript 5.x supported by Deno **_except_** for the Stage 3 decorators :(

**Comment by petamoriken** [comment]
We need to wait for the V8 implementation of Stage 3 Decorators.
https://bugs.chromium.org/p/v8/issues/detail?id=12763

**Comment by lucacasonato** [comment]
I propose we enable JS standard decorators in 1.37 for TS / TSX / JSX via transpilation. They would not yet be supported in plain JS until V8 ships them.

This would be a breaking change. Users of existing Typescript "experimentalDecorators" that work in Deno <1.36 would stop working in Deno >=1.37. They would have to be replaced with JS standard decorators.

I am thinking that this breaking change is OK, because experimental decorators have always been an _experimental_ feature with the expectation of deprecation / removal after ES decorators ship.

If there are major concerns raised regarding this proposal, we could also defer this change as a breaking change for Deno 2.0.

**Comment by jiawei397** [comment]
Today, I took a closer look at the decorators in TypeScript 5+, and I found that they are not quite as I expected. One major issue is that `ParameterDecorator` is currently not supported. If `experimentalDecorators` were to be completely disabled at this point, it would render code like the following, commonly used in frameworks like NestJS, unusable:

```ts
@Get()
async findOne(@User() user: UserEntity) {
  console.log(user);
}
```

If the changes were purely at the syntax level, I would be willing to adapt my framework code. However, this breaking update is quite significant, and I would advise careful consideration. 

It might be a good idea to retain support for the old version of `experimentalDecorators` temporarily. Alternatively, we could wait for TypeScript's official deprecation or work towards implementing `ParameterDecorator`.

The following content is from this [link](https://blog.logrocket.com/using-modern-decorators-typescript/#decorators-typescript-5):

<img width="771" alt="image" src="https://github.com/denoland/deno/assets/19886374/49e78679-f13c-4026-bc3b-3854480a9b2a">

**Comment by nberlette** [comment]
@lucacasonato Could we do something akin to what the TypeScript team has done? What I mean is moving the legacy decorators into a fully legacy stage, **_requiring_** the experimentalDecorators flag to be enabled to use them (which would therefore disable the Stage 3 decorators)? 

IIRC Ron Buckton and crew came to the conclusion that enabling the ESDecorators by default and scrapping the legacy style would be too much of a catastrophic change. Since most users already had the `experimentalDecorators` flag enabled (from all the years in which it was required), however, having that flag be the switch between new / old seemed to be the route with minimal collateral damage.

If Deno were to introduce something along the same lines, allowing users to decide which style to enable on a per-project basis (but with the Stage 3 as default obv), I feel like it would probably go over pretty smoothly with most users... It would _still_ be a breaking change for those who adapted to recent versions in which the decorators were enabled without the flag. 

Sorry for the ramble. I'm passionate about decorators with strong types 🥺

**Comment by pokatomnik** [comment]
Totally agree with @nberlette: the flag could be the most appropriate.
- when new decorators don't work, Deno doesn't fully support Typescript
- when new decoratos work, Deno breaks old approach.
So It would be extremely useful to let the user decide whether they supporting legacy code OR making a new one with perfect shiny well-typed modern decorators:)

**Comment by Ciantic** [comment]
1. It should default to newish standard behavior (Stage 3 decorators), with `experimentalDecorators: false`
2. Optionally you can opt to experimental decorators (old TypeScript only way) with `experimentalDecorators: true`

It could be configured like JSX is configured, at the top of file we now have:
```
/** @jsxImportSource https://esm.sh/preact */
```

To opt-in to legacy decorators
```
/** @experimentalDecorators true */
```

**Comment by greentore** [comment]
I have been grappling with cryptic type errors and unexpected behavior for a while before realizing stage 3 decorators are not supported and non standard `experimentalDecorators` syntax is used instead. They should be disabled by default at the very least.

**Comment by bartlomieju** [comment]
After discussing this offline, we've decided that Stage 3 Decorators proposal will be enabled by default in Deno v1.40, which means that `experimentalDecorators` setting will be switched to `false` by default.

If you are using `experimentalDecorators` syntax, prepare yourself for this change by adding this bit to your `deno.json` file:
```json
{
  "compilerOptions": {
    "experimentalDecorators": true
  }
}
```

**Comment by nberlette** [comment]
@bartlomieju 
> After discussing this offline, we've decided that Stage 3 Decorators proposal will be enabled by default in Deno v1.40, which means that `experimentalDecorators` setting will be switched to `false` by default.
> 
> If you are using `experimentalDecorators` syntax, prepare yourself for this change by adding this bit to your `deno.json` file:
> 
> ```json
> {
>   "compilerOptions": {
>     "experimentalDecorators": true
>   }
> }
> ```

Seeing this reply in my feed just made my day! I've personally been quite eagerly waiting for this since the beginning of the year, so I'd like to thank you and the rest of the Deno team for pulling this one together. 

Am I correct in assuming that enabling the `experimentalDecorators` flag will also disable the Stage 3 decorators, similar to the standard TypeScript v5 behavior? And secondly, will the `emitDecoratorMetadata` flag emulate the standard TypeScript behavior?

Thanks

**Comment by dsherret** [comment]
> Am I correct in assuming that enabling the experimentalDecorators flag will also disable the Stage 3 decorators, similar to the standard TypeScript v5 behavior?

Yes.

> And secondly, will the emitDecoratorMetadata flag emulate the standard TypeScript behavior?

`emitDecoratorMetadata` is an `experimentalDecorator` feature so it will continue working as it does today in Deno with the `experimentalDecorator` feature enabled.

See here how tsc errors without using experimental decorators:

```shellsession
% cat tsconfig.json
{
  "compilerOptions": {
    "experimentalDecorators": false,
    "emitDecoratorMetadata": true
  }
}
% npx tsc          
tsconfig.json:3:5 - error TS5052: Option 'emitDecoratorMetadata' cannot be specified without specifying option 'experimentalDecorators'.

3     "experimentalDecorators": false,
      ~~~~~~~~~~~~~~~~~~~~~~~~

tsconfig.json:4:5 - error TS5052: Option 'emitDecoratorMetadata' cannot be specified without specifying option 'experimentalDecorators'.

4     "emitDecoratorMetadata": true
      ~~~~~~~~~~~~~~~~~~~~~~~


Found 2 errors in the same file, starting at: tsconfig.json:3
```

**Comment by pokatomnik** [comment]
Just tested with my small [repo](https://github.com/pokatomnik/ts-modern-decorators-test). Works as expected. Thanks again, awesome!

**Comment by uptownhr** [comment]
> After

@cleverplatypus Did this issue ever fixed for you? 

I'm running into an issue where decorators are working when ran locally but does not when deployed to Deno cloud. 

https://github.com/denoland/deno/issues/19160#issuecomment-1949548670

**Comment by cleverplatypus** [comment]
Yes @uptownhr, thanks. This was fixed quite some time ago, with 1.41, I believe.

This related issue is still pending and it's not obvious if anybody is paying attention to it, tho: https://github.com/denoland/deploy_feedback/issues/633
